[Remote Inspection] Redundant nearby targets should not be surfaced to Web Inspector

whsieh opened the pull request:
#### 5e8cbc9e70f01a95d48990d9328ca7f167714e41
<pre>
[Remote Inspection] Redundant nearby targets should not be surfaced to Web Inspector
<a href="https://bugs.webkit.org/show_bug.cgi?id=273953">https://bugs.webkit.org/show_bug.cgi?id=273953</a>

Reviewed by Abrar Rahman Protyasha.

Currently, it&apos;s possible for the &quot;nearby targets&quot; heuristic to discover redundant elements, in the
case where a nearby target is a parent of another. In this case, it makes sense to only surface the
parent target. Fix this by implementing an algorithm to filter out such elements from the final list
of nearby targets, such that the final list does not contain any target that is a child of another
target.

Changes covered by augmenting the API test ElementTargeting.NearbyOutOfFlowElements.

* Source/WebCore/page/ElementTargetingController.cpp:
(WebCore::filterRedundantNearbyTargets):
(WebCore::ElementTargetingController::extractTargets):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/element-targeting-2.html:

Canonical link: <a href="https://commits.webkit.org/278583@main">https://commits.webkit.org/278583@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/624af83d1da41eff8758c0ca33275dc60297788f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/50987 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/30286 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/3313 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/54244 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/1676 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/53288 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/49/builds/36606 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/1341 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/41532 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/53085 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/27928 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/43926 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/22652 "Passed tests") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/25265 "Unexpected infrastructure issue, retrying build") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/1181 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/9427 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/47243 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/1255 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/55838 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/26094 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/1155 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/48941 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/27345 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/43997 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/48072 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11164 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/28223 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/27076 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebCore/page/ElementTargetingController.cpp:
```diff
@@ -825,6 +825,38 @@ std::pair<Vector<Ref<Node>>, RefPtr<Element>> ElementTargetingController::findNo
 
+static Vector<Ref<Element>> filterRedundantNearbyTargets(HashSet<Ref<Element>>&& unfilteredNearbyTargets)
+{
+    HashMap<Ref<Element>, bool> shouldKeepCache;
```
Nit: Is it not okay to make this a `HashSet<Ref<Element>>`? Where presence indicates the `true` value for the element key, and absence indicates `false`. That would be a little bit more space efficient.

Basically we would condition adding to the set like such:

```
for (auto& ancestor : ancestorsOfTarget) {
    if (shouldKeep)
        shouldKeepCache.add(ancestor);
}
```

whsieh replied:
Hm.. I don't think this would work, since this cache is currently used like so:
• `true` → keep
• `false` → discard
• (missing key) → not cached yet

whsieh replied:
I considered using 2 `HashSet`s instead (one representing keep and another representing discard), but ultimately went with a single map.

aprotyas replied:
Okay sounds good, I misunderstood the code!

whsieh committed:
[Remote Inspection] Redundant nearby targets should not be surfaced to Web Inspector
https://bugs.webkit.org/show_bug.cgi?id=273953

Reviewed by Abrar Rahman Protyasha.

Currently, it's possible for the "nearby targets" heuristic to discover redundant elements, in the
case where a nearby target is a parent of another. In this case, it makes sense to only surface the
parent target. Fix this by implementing an algorithm to filter out such elements from the final list
of nearby targets, such that the final list does not contain any target that is a child of another
target.

Changes covered by augmenting the API test ElementTargeting.NearbyOutOfFlowElements.

* Source/WebCore/page/ElementTargetingController.cpp:
(WebCore::filterRedundantNearbyTargets):
(WebCore::ElementTargetingController::extractTargets):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/element-targeting-2.html:

Canonical link: https://commits.webkit.org/278583@main
